opnieuw/retries.py

```diff
@@ -68,98 +68,84 @@ def calculate_exponential_multiplier(
     return multiplier
 
 
-class DoCall:
+class BackoffCalculator:
     """
-    An instance which tells the retry decorator to attempt the function.
-    """
-
-
-class DoWait(NamedTuple):
-    """
-    An instance which tells the retry decorator to wait.
+    Class responsible for calculating backoff periods.
 
-    This is used to calculate the sleep time in seconds for in the retry decorator.
-    """
-
-    # Number of calls made so far.
-    attempts_so_far: int
-    # Wait at least this much.
-    min_seconds: float
-    # Wait at most this much.
-    max_seconds: float
-    # Number of seconds left before the user's requested seconds are exceeded
-    seconds_left: float
-
-
-Action = Union[DoCall, DoWait]
-
-
-class RetryState:
-    """
-    Calculate whether to do the function call or to sleep for retrying.
+    Will consider the maximum amount of backoffs and a maximum backoff window.
     """
 
     def __init__(
         self,
         clock: Clock,
-        max_calls_total: int,
-        retry_window_after_first_call_in_seconds: int,
+        max_backoffs_total: int,
+        backoff_window_after_first_call_in_seconds: int,
     ) -> None:
         self.clock = clock
-        self.max_calls_total = max_calls_total
+        self.max_backoffs_total = max_backoffs_total
         self.deadline_second = (
-            self.clock.seconds_since_epoch() + retry_window_after_first_call_in_seconds
+            self.clock.seconds_since_epoch()
+            + backoff_window_after_first_call_in_seconds
         )
-
         self.base_in_seconds = calculate_exponential_multiplier(
-            max_calls_total, retry_window_after_first_call_in_seconds
+            max_backoffs_total, backoff_window_after_first_call_in_seconds
         )
+        self.backoffs = 0
 
-    def __iter__(self) -> Iterator[Action]:
-        for attempt in range(0, self.max_calls_total):
-            # attempt the actual function call
-            yield DoCall()
+    def get_backoff(self) -> float | None:
+        """
+        Return the amount of seconds to backoff.
 
-            wait_seconds = self.base_in_seconds * 2**attempt
-            seconds_left = self.deadline_second - self.clock.seconds_since_epoch()
+        None indicates that there should be no more backoffs.
```

Comment:
It would be good to mention this means the decorator will raise the exception that is being handled.